### ISSUE TITLE
Pyright cannot pick the right overload for some decorator function factories.

### ISSUE THREAD

**Issue by weakish** [microsoft/pyright#728]
Some function decorators supports both bare calls and calls with arguments.
These function decorators can be annotated with `@overload`.
However, Pyright seems not able to pick the right overload.

Example:

```python
from typing import Callable, TypeVar, overload, Optional, Union

F = TypeVar('F', bound=Callable[[], None])

# Bare decorator usage
@overload
def atomic(__func: F) -> F: ...

# Decorator with arguments
@overload
def atomic(*, savepoint: bool = True) -> Callable[[F], F]: ...

# Implementation
def atomic(__func: Optional[Callable[..., None]] = None, *, savepoint: bool = True) -> Union[Callable[[], None], Callable[[Callable[[], None]], Callable[[], None]]]:
    def decorator(func: Callable[..., None]) -> Callable[..., None]:
        ...  # Code goes here
    if __func is not None:
        return decorator(__func)
    else:
        return decorator

# Usage
@atomic
def func1() -> None: ...

@atomic(savepoint=False)
def func2() -> None: ...
```

Pyright (tested with the vscode extension 1.1.42) complains `No parameter named "savepoint"` for `@atomic func2`.
It seems that Pyright picks the bare call version for func2, instead of the with argument version.

`mypy --strict above_code.py` reports no error:

> Success: no issues found in 1 source file

The code sample is adapted from the example given in [mypy documentation][doc].
I made some tweaks to make it passing `mypy --strict`.

[doc]: https://mypy.readthedocs.io/en/stable/generics.html#decorator-factories

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next version of Pyright.

**Comment by erictraut** [comment]
This is fixed in Pyright 1.1.43, which I just published.
